_posts/2023-10-04-#28107.md

```diff
@@ -7,17 +7,55 @@ authors: [dergoegge]
+
+- This PR aims to improve type safety by introducing the new [`transaction_identifier`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R18) type, with `Txid` and `Wtxid` typedefs for brevity.
+
+- There are instances where a variable could be either a `wtxid` or a `txid`. To navigate this, in some places like [`most_recent_block_txs`](https://github.com/bitcoin/bitcoin/blob/6619d6a8dca5a4d8e664a76526ac6bef3eb17831/src/net_processing.cpp#L1916-L1918), the `uint256` type is used. In other contexts such as [`CTxMemPool::info`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/txmempool.cpp#L862), the [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) type is adopted.
+
 ## Questions
+
+### Concept
+
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. What does it mean for a transaction identifier to be type-safe? Why is that important or helpful? Are there any downsides?
+
+1. Conceptually, when writing new code that requires referring to transactions, when should you use `txid` and when should you use `wtxid`?
+
+1. In which concrete way(s) could using `transaction_identifier` instead of `uint256` help find existing bugs? On the other hand, could this change introduce new bugs?
+
+1. SegWit activated at block height 481,824. Do transactions in blocks < 481,824 also have a `wtxid`? For example, what about transaction with `txid` `dc01120a639283c11eeada6ce540178adfbbca7fe0d2f41076367cbc9e901822`?
+
+1. Are there any places where we currently return a (w)txid as a `uint256` where it would not make sense to return a `transaction_identifier` instead?
+
+1. We already have a [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) class. How does it already enforce type correctness, and how is it different from the approach in this PR?
+
+1. Are there any places where `GenTxid` is used but using `transaction_identifier` would be better?
+
+### Code Review
+
+{:start="9"}
+1. How is `transaction_identifier` able to subclass `uint256`, given that, in C++, integers are types and not classes?
+
+1. Why does `transaction_identifier` subclass `uint256` instead of being a completely new type?
+
+1. We define `Txid` and `Wtxid` as typedefs for `transaction_identifier<has_witness>`. Do you see any reason for code to directly use the `transaction_identifier` type? Should we forbid that, and if so, how?
```

Comment:
I slightly prefer this kind of voice for talking about the PR code, but just a nit
```suggestion
1. `Txid` and `Wtxid` are defined as typedefs for `transaction_identifier<has_witness>`. Do you see any reason for code to directly use the `transaction_identifier` type? Should we forbid that, and if so, how?
```

Reply:
Agreed, done, for all other instances too.